### PR TITLE
Formalise guidelines around writing tests

### PR DESCRIPTION
**Pull request by novocaine**
As discussed internally and in #element-dev

---
This PR currently has no changelog labels, so will not be included in changelogs.

A reviewer can add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is, or add `Type: [enhancement/defect/task]` to the description and I'll add them for you.

### CODE REVIEW

**Review by kerryarchibald — APPROVED** [review]
LGTM

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -100,18 +100,34 @@ checks, so please check back after a few minutes.
...
+
+For new user facing features in `matrix-react-sdk` or `element-web`, you
+must include:
+
+1. Comprehensive unit tests written in Jest. These are located in `/test`.
+2. "happy path" end-to-end tests.
+   These are located in `/test/end-to-end-tests`. Ideally, you would also
+   include tests for edge and error cases.
+
+For bugs in those repos, your change must include at least one unit tests or
```
Comment (andybalaam) [inline CONTRIBUTING.md:113]:
```suggestion
For bugs in those repos, your change must include at least one unit test or
```

**Review by andybalaam — APPROVED** [review]
One typo - otherwise wholeheartedly agree!

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -100,18 +100,34 @@ checks, so please check back after a few minutes.
...
-percentage for the diff.
+Your PR should include tests.
+
+For new user facing features in `matrix-react-sdk` or `element-web`, you
+must include:
+
+1. Comprehensive unit tests written in Jest. These are located in `/test`.
+2. "happy path" end-to-end tests.
+   These are located in `/test/end-to-end-tests`. Ideally, you would also
+   include tests for edge and error cases.
```
Comment (t3chguy) [inline CONTRIBUTING.md:111]: this only applies to react-sdk

**Review by t3chguy — APPROVED** [review]

File: CONTRIBUTING.md (hunk 2 above)
Comment (turt2live) [inline CONTRIBUTING.md:111]: we'll need to fix the test suite to actually run easily on all environments to make this enforceable.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -100,18 +100,34 @@ checks, so please check back after a few minutes.
...
-
-We don't require tests for bug fixes (T-Defect) but strongly encourage regression
-tests for the bug itself wherever possible.
-
-In the future we may formalise this more with a minimum test coverage
-percentage for the diff.
+Your PR should include tests.
+
+For new user facing features in `matrix-react-sdk` or `element-web`, you
+must include:
```
Comment (turt2live) [inline CONTRIBUTING.md:106]: does this include features behind labs? Not sure where the clause went, but previously we had a (potentially informal) policy that tests had to be written for a feature to leave labs, but not to enter labs. I'd like to keep this clause in place.

**Review by turt2live — CHANGES_REQUESTED** [review]
mostly wording clarifications - the principle is agreeable.

File: CONTRIBUTING.md (hunk 3 above)
Comment (novocaine) [reply CONTRIBUTING.md:106]: We can discuss, but we want to shoot for writing unit tests alongside new work, not after the fact; this means the code is much more likely to be testable and you get the benefit of having the tests there to make sure you're not breaking stuff as you go.

For e2e tests, I can see the argument for adding them at the end of feature development.

File: CONTRIBUTING.md (hunk 2 above)
Comment (novocaine) [reply CONTRIBUTING.md:111]: Do we have issues tracking the current problems?

**Review by gsouquet — APPROVED** [review]

File: CONTRIBUTING.md (hunk 3 above)
Comment (jryans) [reply CONTRIBUTING.md:106]: The previous policy @turt2live mentions is described in https://github.com/vector-im/element-web/blob/develop/docs/review.md#code-quality, which I guess partially overlaps with this file (and perhaps conflicts with it).

Not sure which repo should be the canonical home of team process docs, but Element Web seems a tad more logical to me... 🤷 There are other useful tidbits in https://github.com/vector-im/element-web/blob/develop/docs/review.md, so it would seem best to preserve and update it where needed.

Comment (novocaine) [reply CONTRIBUTING.md:106]: Thanks. That policy isn't linked to from the element-web README afaict, which links to the develop guide for matrix-react-sdk, which links to this file.

Comment (jryans) [reply CONTRIBUTING.md:106]: Yes, but all of them are documents we can edit... 😜 

Perhaps good to add separate document clean up task. At the moment, there are enough separate documents in different repos it's likely overwhelming for someone new (contributor or Element employee alike), especially when there are several covering similar topics with different text.

**Review by dbkr — APPROVED** [review]
Seems reasonable! Some good points about the various different docs at different levels of the tree: we should try & clean these up a bit (I think I'm convinced that the right place for things to live would be element-web).

File: CONTRIBUTING.md (hunk 2 above)
Comment (turt2live) [reply CONTRIBUTING.md:111]: not off the top of my head, but the problems seem easily encountered. We also discussed a few of them in other channels.

It's very possible that it's just a situation of us needing to try installing them again on affected platforms and see how it goes (and report bugs).

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -110,6 +110,13 @@ must include:
    These are located in `/test/end-to-end-tests`. Ideally, you would also
    include tests for edge and error cases.
 
+Unit tests are expected even when the feature is in labs. It's good practice
```
[the remark this answers is not in the document]
Comment (novocaine) [reply CONTRIBUTING.md:113]: I think if we're saying some code can escape the threshold of typical quality, we need a more powerful distinction for marking it as "experimental" than just labs. A lot of labs code these days really is intended for eventual production, and skipping tests during development is problematic because adding tests later usually doesn't work - the code won't be very testable and will need reworking.

**Review by SimonBrandner — APPROVED** [review]
